File: tests/DataProperty_1.cc
// -*- lsst-c++ -*-
#include "lsst/fw/DataProperty.h"
#include <lsst/fw/Citizen.h>
#include <lsst/fw/Trace.h>

using namespace std;
using namespace lsst::fw;

class Foo {
     int gurp;
     std::string murp;
     double durp;
};

void test() {
    DataPropertyPtrT root(new DataProperty("root"));

     DataPropertyPtrT prop1(new DataProperty("name1", std::string("value1")));
     DataPropertyPtrT prop2(new DataProperty("name2", 2));
     DataPropertyPtrT prop2a(new DataProperty("name2", 4));
     

     root->addProperty(prop1);
     root->addProperty(prop2);

     Foo foo1;
     DataPropertyPtrT prop3(new DataProperty("name3", foo1));
     root->addProperty(prop3);

     root->addProperty(prop2a);

//      DataPropertyPtrT dpPtr = root->find("name2");
//      dpPtr->print("\t");

//      // check find without reset to beginning
//      dpPtr = root->find("name2", false);
//      dpPtr->print("\t");

//      dpPtr = root->find("name1");
//      dpPtr->print("\t");
//      dpPtr = root->find("name3");
//      dpPtr->print("\t");

     // Try nested property list
     
     DataPropertyPtrT nested(new DataProperty("nested"));

     DataPropertyPtrT nprop1(new DataProperty("name1n", std::string("value1")));
     DataPropertyPtrT nprop2(new DataProperty("name2n", 2));
     

     nested->addProperty(nprop1);
     nested->addProperty(nprop2);

     root->addProperty(nested);

     root->print("\t");

     // Check copy constructor

     DataPropertyPtrT rootCopy(new DataProperty(*root));

     // Explicitly destroy root

     root.reset();

     std::cout << "Explicit destruction done" << std::endl;

     // Check that rootCopy is still OK...

     rootCopy->print("\t");
     
}     

void test2()
{
    boost::any foo = stringToAny("-1234");
    boost::any foo2 = stringToAny("1.234e-1");
    boost::any foo3 = stringToAny("'This is a Fits string'");
    
    DataPropertyPtrT fooProp(new DataProperty("foo", foo));
    DataPropertyPtrT fooProp2(new DataProperty("foo2", foo2));
    DataPropertyPtrT fooProp3(new DataProperty("foo3", foo3));
    fooProp->print();
    fooProp2->print();
    fooProp3->print();
}

int main() {
    Trace::setVerbosity("fw.DataProperty", 10);

    test();
     //
     // Check for memory leaks
     //
     if (Citizen::census(0) == 0) {
         cerr << "No leaks detected" << endl;
     } else {
         cerr << "Leaked memory blocks:" << endl;
         Citizen::census(cerr);
     }

     test2();
}

File: tests/DataProperty_1.py
import pdb                              # we may want to say pdb.set_trace()
import unittest
import lsst.fw.Core.tests as tests
import lsst.fw.Core.fwLib as fw

try:
    type(verbose)
except NameError:
    verbose = 0
    fw.Trace_setVerbosity("fw.DataProperty", verbose)

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

class DataPropertyTestCase(unittest.TestCase):
    """A test case for DataProperty"""
    def setUp(self):
        self.root = fw.DataProperty("root")

        self.values = {}; props = []
        
        n = "name1"; self.values[n] = "value1"
        props += [fw.DataProperty(n, self.values[n])]

        n = "name2"; self.values[n] = 2
        props += [fw.DataProperty(n, self.values[n])]
        props += [fw.DataProperty(n, 2*self.values[n])] # add with different value

        n = "name3"
        if False:                           # this won't work, as I don't
                                            # have boost::any working from python
            class Foo:
                __slots__ = ["gurp", "murp", "durp"]

            self.values[n] = Foo()
        else:
            self.values[n] = "Foo()"
        props += [fw.DataProperty(n, self.values[n])]

        for prop in props:
            self.root.addProperty(prop)

    def tearDown(self):
        del self.root
        self.root = None

    def testName1(self):
        """Check "name1", a string valued DataProperty"""
        
        n = "name1"
        dpPtr = self.root.find(n)
        assert dpPtr.get() != None, "Failed to find %s" % n
        self.assertEqual(dpPtr.getValueString(), self.values[n])

    def testDataPtrType(self):
        """Check that the getValueXXX routines get the right types"""
        dpPtr = self.root.find("name1")
        self.assertRaises(RuntimeError, dpPtr.getValueInt)

        dpPtr = self.root.find("name2")
        self.assertRaises(RuntimeError, dpPtr.getValueString)

    def testName2(self):
        """Check "name2", an int valued DataProperty with two definitions"""
        
        n = "name2"

        dpPtr = self.root.find(n)
        assert dpPtr.get() != None, "Failed to find %s" % n
        assert dpPtr.getValueInt() == self.values[n]

        dpPtr = self.root.find(n, False)
        assert dpPtr.get() != None, "Failed to find %s" % n
        assert dpPtr.getValueInt() == 2*self.values[n]

        dpPtr = self.root.find(n, False)
        assert dpPtr.get() == None, "Failed to find %s" % n

        dpPtr = self.root.find(n)
        self.assertEqual(dpPtr.getValueInt(), self.values[n])

    def testName3(self):
        """Check name3, which (should have) a non-{int,string} type"""
        n = "name3"
        dpPtr = self.root.find(n)
        assert dpPtr.get() != None, "Failed to find %s" % n

    def testUndefined(self):
        """Check that we can't find a data property that isn't defined"""
        dpPtr = self.root.find("undefined")
        assert dpPtr.get() == None, "Found non-existent DataProperty"

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

class NestedDataPropertyTestCase(unittest.TestCase):
    """A test case for nested DataProperty"""
    def setUp(self):
        self.root = fw.DataProperty("root")

        nested = fw.DataProperty("nested")

        self.values = {}; props = []
           
        n = "name1"; self.values[n] = "value1"
        props += [fw.DataProperty(n, self.values[n])]

        n = "name2"; self.values[n] = 2
        props += [fw.DataProperty(n, self.values[n])]

        for prop in props:
            nested.addProperty(prop)

        self.root.addProperty(nested)
        
    def tearDown(self):
        del self.root
        self.root = None

    def testCopyConstructor(self):
        """Check copy constructor"""
    
        rootCopy = fw.DataProperty(self.root)

        # Explicitly destroy root
        del self.root; self.root = None
    
        # Check that rootCopy is still OK...
        assert rootCopy.repr() != None, "rootCopy is mangled"

    def testNested(self):
        """Extract root node"""
        contents = self.root.getContents()
        self.assertEqual(len(contents), 1)
        self.assertEqual(contents[0].getName(), "nested")

    def testNested2(self):
        """Extract nested contents"""
        contents = self.root.getContents()

        for n in ("name1", "name2"):
            dpPtr = contents[0].find(n)
            assert dpPtr.get() != None, "failed to find %s" % n
            getValue = (dpPtr.getValueString if (n == "name1") else dpPtr.getValueInt)
            self.assertEqual(getValue(), self.values[n])

    def testRegex(self):
        """Find DataProperty using boost::regex"""
        contents = self.root.getContents()

        while True:
            dpPtr = contents[0].match("^name[0-9]+", False)
            if not dpPtr.get():
                break
            n = dpPtr.getName()

            getValue = (dpPtr.getValueString if (n == "name1") else dpPtr.getValueInt)
            self.assertEqual(getValue(), self.values[n])

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

def suite():
    """Returns a suite containing all the test cases in this module."""
    tests.init()

    suites = []
    suites += unittest.makeSuite(DataPropertyTestCase)
    suites += unittest.makeSuite(NestedDataPropertyTestCase)
    suites += unittest.makeSuite(tests.MemoryTestCase)
    return unittest.TestSuite(suites)

if __name__ == "__main__":
    unittest.main()
